### DISCUSSION TITLE
Best practices on using historical data API for a GPS trace #803

### DISCUSSION THREAD

**Discussion by sarthak** [open-meteo/open-meteo#804] [Q&A] [answered]
Hello, I am interested in using historical data API to evaluate some weather attributes along a GPS trace. A GPS trace is just time-series GPS data. The traces that I have are quite large that span multiple years, and have a very high sampling rate (1 point every few seconds) collectively totalling to multi-million points. Obviously it is neither feasible nor necessary to feed every GPS point into the API as most points lying close together in both temporal and spatial dimensions can be approximated to have the same weather attributes. I am interested in understanding when to consider two time-series GPS points "different" enough to make a new API call. For example, given that historical weather models have temporal granularity of 1 hour, two GPS points different more than one hour in the time dimension may be considered "different". Is there any rule-of-thumb / analytical formula that we can use for spatial dimensions as well? Any recommendations on best practices for such a use case are highly appreciated.

**Comment by patrick-zippenfenig** [comment]
Hi,
weather typically does not change that noticeable every 5 minutes / 1 km. Especially the historical data API only has a resolution of ~9 km. 

You can use the [bounding box feature](https://github.com/open-meteo/open-meteo/pull/734) (mostly undocumented on not available for all domains). For example it does not work with the historical API except is you limit yourself to ERA5-Land at (10km) OR ERA5 (25 km). You can also use the upcoming[ historical forecast API](https://open-meteo.com/en/docs/historical-forecast-api#start_date=2024-05-06&end_date=2024-05-20) with a bounding box selecting the model DWD ICON GLOBAL or any local domain, depending on your region. Best match does not work.

**Reply by patrick-zippenfenig** [reply]
@sarthak did you manage to get it working?

**Reply by sarthak** [reply]
Hi @patrick-zippenfenig thanks for prompt response here, but I am not sure I completely follow. It seems that bounding box would return individual grid cells, but then how does that help me determine the weather at a particular GPS? I think that would require me to do some kind of interpolations which I am not sure about, especially if elevation is variable.

I also did not follow your intuition on why a bounding box would be a better strategy for getting weather data along a very large GPS trace (large in terms of total distance spanned by the trace) versus just downsampling the GPS trace altogether and then making individual API calls per point in the trace?

**Reply by patrick-zippenfenig** [reply]
Yes, afterwards you need to interpolate (or take the nearest neighbour). I would load all data into a pandas data frame and look for options to interpolate.

**Reply by sarthak** [reply]
I have been reading your blog posts (very interesting btw) on the topic and I have realised (specifically through [this](https://openmeteo.substack.com/p/improving-weather-forecasts-with) blog) that interpolation between grid cells may not be as easy or primitive to do myself, and I do not want to undo the work you have helpfully done already in this regard which is why I was a bit hesitant.

**Reply by patrick-zippenfenig** [reply, marked as answer]
Good point. Yes, this approach does not consider terrain and does not apply statistical temperature downscaling. Those feature could be implemented in Python, but won't be as easy as using Pandas data frame interpolation.

Just for reference: The API also supports getting data from multiple coordinates at different hours. E.g.:
[...?latitude=**52.52,40.12**&longitude=**13.41,8.1**&start_hour=**2024-02-03T05:00,2024-03-05T05:00**&end_hour=**2024-02-03T05:00,2024-03-05T05:00**&hourly=temperature_2m](https://archive-api.open-meteo.com/v1/archive?latitude=52.52,40.12&longitude=13.41,8.1&start_hour=2024-02-03T05:00,2024-03-05T05:00&end_hour=2024-02-03T05:00,2024-03-05T05:00&hourly=temperature_2m). Note: This call is weighted as 2 calls (= 2 coordinates). It is a fairly inefficient to only fetch single values, because each individual value is still stored in a chunks of compressed data. Typically around 10 locations by 60 time steps. Parameters can also be submitted in POST bodies to work around URL length issues.

**Reply by sarthak** [reply]
> Parameters can also be submitted in POST bodies to work around URL length issues.

Hi, could you give an example for this? I tried some obvious interpretations but they do not seem to be recognised by the server. I am hitting request URI limits and getting 414 for some test batch requests.

**Reply by patrick-zippenfenig** [reply]
If you are using the Python SDK, you can simply pass `method = "POST"`:

```python
url = "https://historical-forecast-api.open-meteo.com/v1/forecast"
params = {
	"latitude": 52.52,
	"longitude": 13.41,
	"start_date": "2024-06-23",
	"end_date": "2024-07-06",
	"hourly": "temperature_2m"
}
responses = openmeteo.weather_api(url, params=params, method="POST")
```

Using curl it looks like this:
```bash
 curl -d "latitude=52.52&longitude=13.41&hourly=temperature_2m" -X POST https://api.open-meteo.com/v1/forecast
```
